cylc/flow/network/resolvers.py

```diff
@@ -329,14 +329,34 @@ async def get_workflows(self, args):
             args)
 
     # nodes
+    def get_node_state(self, node, node_type):
+        """Return state, from node or data-store."""
+        if node_type in DEF_TYPES:
+            return None
+        # Don't lookup data-store state in getattr, as it might not be there.
+        # (args are evaluated first)
+        # If node has no state, it must be in the data-store.
+        with suppress(Exception):
+            return (
+                getattr(node, 'state', None)
+                or self.data_store_mgr.data[
+                    Tokens(node.id).workflow_id
+                ][node_type][node.id].state
```

Comment:
Could you explain what these two different methods mean?

* The first is getting the state from the node, the node as it exists in the data store?
* The second is getting the state from the data store directly. How does this differ from the node?

Reply:
> The first is getting the state from the node, the node as it exists in the data store?

`node` could be from a data-store or a delta-store.

> The second is getting the state from the data store directly. How does this differ from the node?

Yes, if the first returns `None`  or empty string, fetch the state from the data-store... 

If we want only deltas of specified state(s), then we need to filter out those deltas that don't contain state.

Reply:
Ok, that makes sense, is this right?:

```python
            return (
                # try to retrieve the state from the node (could be a delta-store node)
                getattr(node, 'state', None)
                # fall back to the data store (because this could be a delta-store
                # node which might not have the state field set)
                or self.data_store_mgr.data[
                    Tokens(node.id).workflow_id
                ][node_type][node.id].state
```

Reply:
Correct 👍